### PR TITLE
Update JDK and JavaFX

### LINKED ISSUE #8127: [Linux] Keyboard shortcuts for increasing/decreasing font size are not working
On my Ubuntu 20.04 the keyboard shortcuts for increasing/decreasing fonts are not working. Other shortcuts are working. On Windows the Ctrl++ and Ctrl+- shortcuts are working. From menu it is working on Windows and Ubuntu.

The Linux keyboard mapping file from [this manual](https://defold.com/manuals/editor-keyboard-shortcuts/) is not having any shortcut for increasing/decreasing font size (nor a Windows file).

**Expected behavior (REQUIRED):**
Shortcuts are working.

**Defold version (REQUIRED):**
  - Defold 1.6.0

**Platforms (REQUIRED):**
 - Linux (Ubuntu 20.04)

**Workaround (OPTIONAL):**
It is possible to increase and decrease font size using options from menus: View->Increase Font Size and View->Decrease Font Size

**Additional context (OPTIONAL):**

<hr/>

<table>
<tr><td>Build time</td><td>2023-10-06T10:00:24.855519</td></tr>
<tr><td>Defold channel</td><td>editor-alpha</td></tr>
<tr><td>Defold editor sha</td><td>3584a3360b63a831529a2df7de0cb0c715374cc0</td></tr>
<tr><td>Defold engine sha</td><td>d9e9c49ab946c058f29a8b688c862d70f30e9c43</td></tr>
<tr><td>Defold version</td><td>1.6.0</td></tr>
<tr><td>GPU</td><td>Mesa Intel(R) UHD Graphics (CML GT2)</td></tr>
<tr><td>GPU Driver</td><td>2.1 Mesa 21.2.6</td></tr>
<tr><td>Java version</td><td>17.0.5+8</td></tr>
<tr><td>OS arch</td><td>amd64</td></tr>
<tr><td>OS name</td><td>Linux</td></tr>
<tr><td>OS version</td><td>5.15.0-86-generic</td></tr>
<table>

### LINKED ISSUE #8629: Defold editor markdown file: can't see selected text
Since Defold 1.6.3 if I open an .MD file in the editor and try to select a word, a line or a paragraph dragging the mouse I can't see the selection that I made, the text  became highlighted only if I then scroll the page. 

1. Open a tutorial (e.g. RPG map sample)
2. Open README.md
3. Try to select some text with the mouse
4. You can't see the highlighted selection
5. If you then scroll the page up or down with keyboard arrows or with mouse wheel the highlighted selection is visible

Selection of text in markdown file was fine in 1.6.2 release

Defold version 
  - Version 1.6.3 - 1.6.4

 - OS: Linux Mint 21.3 Cinnamon, kernel 5.15.0-97-generic
 - Nvidia Driver 535.161.07

### LINKED ISSUE #8921: Defold adds a space after typing " when using a keyboard layout with dead keys on Windows 10/11
**Describe the bug (REQUIRED)**
When typing any kind of dead key on a keyboard layout that supports it (e.g.`"`, `'`) if you press space after typing one character, it adds a space after adding the character. The correct behaviour should be just to add the character.

**To Reproduce (REQUIRED)**
1. Have Windows installed
2. Use a keyboard layout that supports dead keys (in my case it is ENGLISH INTERNATIONAL)
3. Try to type any dead key, like double quotes, for example, and press space
4. See an extra space being added

**Expected behavior (REQUIRED)**
It should not add the extra space

Right now, this happens: `" `
What should happen is: `"`

**Defold version (REQUIRED):**
  - Version 1.8.0

**Platforms (REQUIRED):**
 - Platforms: N/A
 - OS: Windows
 - Device: N/A

**Screenshots:**

https://github.com/defold/defold/assets/6012227/bfe52b78-cc26-474f-92a5-dac56f12b9e0

### LINKED ISSUE #9704: Update JDK and JavaFX
**Task (REQUIRED):**
For now, we use JDK 17, but to be able to use latest JavaFX we have to use JDK 21.
Bob also should be compatible with JDK 21.

**Expected outcome (REQUIRED):**
Using of the latest JavaFX to fix issues we have:
https://github.com/defold/defold/issues/8921
https://github.com/defold/defold/issues/8629
https://github.com/defold/defold/issues/8127

### PR DESCRIPTION
**Pull request by vlaaad**
Update editor and bob JDKs to 21.
⚠️ While no action is required for the typical use of Defold (i.e. building and bundling from the editor), workflows that use the command line `bob.jar` need to update the JDK version to 21.

Fixes #8921
Fixes #8629
Fixes #8127
Fixes #9704

### CODE REVIEW

File: README_SETUP.md [hunk 1]
```diff
@@ -37,9 +37,9 @@ choco install <package_name>
 ```
 </p></details>
 
-### Required Software - Java JDK 17
+### Required Software - Java JDK 21
 
-Download and install the latest JDK 17 (17.0.5 or later) release from either of these locations:
+Download and install the latest JDK 21 (21.0.5 or later) release from either of these locations:
 
 * [Adoptium/Temurin](https://github.com/adoptium/temurin17-binaries/releases) - The Adoptium Working Group promotes and supports high-quality runtimes and associated technology for use across the Java ecosystem
```
Comment (AGulev) [inline README_SETUP.md:44]: wrong link, it should be 
`https://github.com/adoptium/temurin21-binaries/releases`

File: README_SETUP.md [hunk 2]
```diff
@@ -37,9 +37,9 @@ choco install <package_name>
...
 * [Microsoft OpenJDK builds](https://docs.microsoft.com/en-us/java/openjdk/download#openjdk-17) - The Microsoft Build of OpenJDK is a no-cost distribution of OpenJDK that's open source and available for free for anyone to deploy anywhere
```
Comment (AGulev) [inline README_SETUP.md:45]: The same 
`https://learn.microsoft.com/en-us/java/openjdk/download#openjdk-21`

**Comment by vlaaad** [comment]
https://issuetracker.google.com/issues/365066515

File: editor/jlink-options [hunk 3]
```diff
@@ -19,4 +19,4 @@
 --no-man-pages
 --dedup-legal-notices=error-if-not-same-content
 --no-header-files
---compress=2 # 0 - no compression, 1 - constant string sharing, 2 - zip
+--compress=zip-9 # zip-0 (no compression) to zip-9 (max compression)
```
Comment (vlaaad) [inline editor/jlink-options:22]: jlink renamed options.

File: editor/src/clj/editor/code/syntax.clj [hunk 4]
```diff
@@ -43,7 +43,7 @@
 (defn- replace-back-references
   ^Pattern [^Pattern re ^MatchResult input]
   (re-pattern (reduce (fn [re-string group]
-                        (if-some [capture (.group input group)]
+                        (if-some [capture (.group input ^int group)]
```
Comment (vlaaad) [inline editor/src/clj/editor/code/syntax.clj:46]: Fixed reflection warning: A string arity was introduced.

File: engine/dlib/src/test/wscript [hunk 5]
```diff
@@ -103,7 +103,7 @@ def build(bld):
         name='test_http_server',
         source_root='http_server')
 
-    bld.env["JAVACFLAGS"] = '-g -source 1.7 -target 1.7'.split()
+    bld.env["JAVACFLAGS"] = '-g -source 1.8 -target 1.8'.split()
```
Comment (vlaaad) [inline engine/dlib/src/test/wscript:106]: 1.7 no longer supported

File: engine/gamesys/src/java/com/defold/window/WindowJNI.java [hunk 6]
```diff
@@ -26,21 +26,11 @@ public WindowJNI(Activity activity) {
     }
 
     public void enableScreenDimming() {
-        this.activity.runOnUiThread(new Runnable() {
-            @Override
-            public void run() {
-                activity.getWindow().clearFlags(android.view.WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);
-            }
-        });
+        this.activity.runOnUiThread(() -> activity.getWindow().clearFlags(android.view.WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON));
```
Comment (vlaaad) [inline engine/gamesys/src/java/com/defold/window/WindowJNI.java:29]: somehow an anonymous class failed the build, but a lambda didn't. There is also a warning `Type java.lang.Runnable was not found`, probably related

**Review by matgis — APPROVED** [review]
Looks good to me! ❤️

**Review by AGulev — APPROVED** [review]
🔥

Have you tested the Android build (built and run on a one-sprite project to ensure it builds and runs correctly)?

**Comment by vlaaad** [comment]
I tested an Android build with a spine scene, though I haven't tried using screen dimming or sounds.

**Comment by AGulev** [comment]
Last time we update JDK (1.4.8) it was a breaking change with major version update. Not sure if we have to do it now as well.

**Review by britzl — APPROVED** [review]

**Review by ekharkunov — APPROVED** [review]
👍

**Comment by vlaaad** [comment]
Using the new `bob.jar` with an old JDK will result in an error:
```
Error: LinkageError occurred while loading main class com.dynamo.bob.Bob
        java.lang.UnsupportedClassVersionError: com/dynamo/bob/Bob has been compiled by a more recent version of the Java Runtime (class file version 65.0), this version of the Java Runtime only recognizes class file versions up to 61.0
```

**Pull request merged**